code/modules/mob/abstract/observer/observer.dm

```diff
@@ -159,6 +159,45 @@ Works together with spawning an observer, noted above.
 	if(medHUD)
 		process_medHUD(src)
 
+	teleport_if_needed()
+
+/mob/abstract/observer/proc/on_restricted_level()
+	//Check if they are a staff member
+	if(client && client.holder && (client.holder.rights & (R_MOD|R_ADMIN|R_DEV)))
+		return FALSE
+	
+	//Check if the z level is in the restricted list
+	if (!(z in current_map.restricted_levels))
+		return FALSE
+	
+	return TRUE
+
+/mob/abstract/observer/proc/teleport_to_spawn(var/message)
+	if(!message)
+		message = "You can not freely observe on this z-level."
+
+	//Teleport them back to the ghost spawn
+	var/obj/O = locate("landmark*Observer-Start")
+	if(istype(O))
+		to_chat(src, "<span class='notice'>[message]</span>")
+		forceMove(O.loc)
+
+//Teleports the observer away from z-levels they shouldnt be on, if needed.
+/mob/abstract/observer/proc/teleport_if_needed()
+	if(!on_restricted_level())
+		return
+
+	if(following)
+		if(!istype(following,/mob/living/carbon)) //If they are following something other than a carbon mob, teleport them

```

Comment:
Another point. Maybe add a stop_following call here? Unless this is handled by forcemoving the ghost.